x/crosschain/keeper/msg_server_refund_aborted_tx.go

```diff
@@ -0,0 +1,82 @@
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// check if authorized
+	if msg.Creator != k.zetaObserverKeeper.GetParams(ctx).GetAdminPolicyAccount(observertypes.Policy_Type_group2) {
+		return nil, observertypes.ErrNotAuthorized
+	}
+
+	// check if the cctx exists
+	cctx, found := k.GetCrossChainTx(ctx, msg.CctxIndex)
+	if !found {
+		return nil, types.ErrCannotFindCctx
+	}
+	// make sure separate refund address is provided for bitcoin chain as we cannot refund to tx origin or sender in this case
+	if common.IsBitcoinChain(cctx.InboundTxParams.SenderChainId) && msg.RefundAddress == "" {
+		return nil, errorsmod.Wrap(types.ErrInvalidAddress, "refund address is required for bitcoin chain")
+	}
+
+	// check if the cctx is aborted
+	if cctx.CctxStatus.Status != types.CctxStatus_Aborted {
+		return nil, errorsmod.Wrap(types.ErrInvalidStatus, "CCTX is not aborted")
+	}
+	// check if the cctx is not refunded
+	if cctx.CctxStatus.IsAbortRefunded {
+		return nil, errorsmod.Wrap(types.ErrUnableProcessRefund, "CCTX is already refunded")
+	}
+
+	// Set the proper refund address.
+	// For BTC sender chain the refund address is the one provided in the message in the RefundAddress field.
+	// For EVM chain with coin type ERC20 the refund address is the sender , but can be overridden by the RefundAddress field in the message.
+	// For EVM chain with coin type Zeta the refund address is the tx origin, but can be overridden by the RefundAddress field in the message.
+	// For EVM chain with coin type Gas the refund address is the tx origin, but can be overridden by the RefundAddress field in the message.
+
+	refundAddress := ethcommon.HexToAddress(cctx.InboundTxParams.TxOrigin)
+	if cctx.InboundTxParams.CoinType == common.CoinType_ERC20 {
+		refundAddress = ethcommon.HexToAddress(cctx.InboundTxParams.Sender)
+	}
+	if msg.RefundAddress != "" {
+		if !ethcommon.IsHexAddress(msg.RefundAddress) {
+			return nil, errorsmod.Wrap(types.ErrInvalidAddress, "invalid refund address provided")
+		}
+		refundAddress = ethcommon.HexToAddress(msg.RefundAddress)
+	}
+	// Double check to make sure the refund address is valid
+	if refundAddress == (ethcommon.Address{}) {
+		return nil, errorsmod.Wrap(types.ErrInvalidAddress, "invalid refund address")
+	}
+	// Check if aborted amount is available to maintain zeta accounting
+	// NOTE: Need to verify if this check works / is required in athens 3
+	if cctx.InboundTxParams.CoinType == common.CoinType_Zeta {
+		err := k.RemoveZetaAbortedAmount(ctx, cctx.InboundTxParams.Amount)
+		if err != nil {
+			return nil, errorsmod.Wrap(types.ErrUnableProcessRefund, err.Error())
+		}
+	}
+	// refund the amount
+	tmpCtx, commit := ctx.CacheContext()
+	err := k.RefundAbortedAmountOnZetaChain(tmpCtx, cctx, refundAddress)
+	if err != nil {
+		return nil, errorsmod.Wrap(types.ErrUnableProcessRefund, err.Error())
+	}
+	commit()
```

Comment:
I agree , But it us safer since we are calling zEVM inside , unless you can think of a reason this might cause problems , I would rather keep it

Reply:
@lumtis : if you CallEVM you generally need to be in a temporary context. 
otherwise if the enclosing cosmos tx reverts, the EVM state may *NOT* revert, which is extremely dangerous.

Reply:
BTW @kingpinXD I prefer to move commit() the end, right before the return (success).

Reply:
The two statements after the commit , do not return errors , and so  we would always return success after this line. 
I think keeping the commit here looks better visually , as we are only wrapping the refund in a temp context

Reply:
Ok let's keep it that way